File: Design-Patterns-Using-Ruby/Command/CommandPattern.rb
# Invoker
class Invoker
    def initialize
        @commands = []
    end
  
    def set_command(command)
        @commands << command
    end
  
    def execute_commands
        @commands.each(&:execute)
    end
  
    def unexecute_commands
        @commands.each(&:unexecute)
    end
end

# Command
class Command
    def execute
        raise NotImplementedError, 'Subclasses must implement this method'
    end
  
    def unexecute
        raise NotImplementedError, 'Subclasses must implement this method'
    end
end

# ConcreteCommand
class ConcreteCommand < Command
    def initialize(receiver)
        @receiver = receiver
    end
  
    def execute
        @receiver.action('Action 1')
    end
  
    def unexecute
        @receiver.action('Action 2')
    end
end

# Receiver
class Receiver
    def action(action)
        puts action
    end
end

# Client Code
receiver = Receiver.new
concrete_command = ConcreteCommand.new(receiver)
invoker = Invoker.new

invoker.set_command(concrete_command)
invoker.execute_commands
invoker.unexecute_commands

=begin 
Action 1
Action 2
=end

File: Design-Patterns-Using-Go/Command/CommandPattern.go
package main

import "fmt"

// Command interface
type Command interface {
	execute()
	unexecute()
}

// Invoker struct
type Invoker struct {
	commands []Command
}

// setCommand method for Invoker
func (i *Invoker) setCommand(command Command) {
	i.commands = append(i.commands, command)
}

// executeCommands method for Invoker
func (i *Invoker) executeCommands() {
	for _, command := range i.commands {
		command.execute()
	}
}

// unexecuteCommands method for Invoker
func (i *Invoker) unexecuteCommands() {
	for _, command := range i.commands {
		command.unexecute()
	}
}

// ConcreteCommand struct
type ConcreteCommand struct {
	receiver Receiver
}

// NewConcreteCommand constructor
func NewConcreteCommand(receiver Receiver) *ConcreteCommand {
	return &ConcreteCommand{
		receiver: receiver,
	}
}

// execute method for ConcreteCommand
func (c *ConcreteCommand) execute() {
	c.receiver.action("Action 1")
}

// unexecute method for ConcreteCommand
func (c *ConcreteCommand) unexecute() {
	c.receiver.action("Action 2")
}

// Receiver struct
type Receiver struct{}

// action method for Receiver
func (r *Receiver) action(action string) {
	fmt.Println(action)
}

// Client Code
func main() {
	receiver := &Receiver{}
	concreteCommand := NewConcreteCommand(*receiver)
	invoker := &Invoker{}
	invoker.setCommand(concreteCommand)
	invoker.executeCommands()
	invoker.unexecuteCommands()
}

/*
Action 1
Action 2
*/

